flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayExceptFunction.java

```diff
@@ -0,0 +1,134 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_EXCEPT}. */
+@Internal
+public class ArrayExceptFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+    private final EqualityAndHashcodeProvider equalityAndHashcodeProvider;
+
+    public ArrayExceptFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_EXCEPT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        this.equalityAndHashcodeProvider = new EqualityAndHashcodeProvider(context, dataType);
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        equalityAndHashcodeProvider.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData arrayOne, ArrayData arrayTwo) {
+        try {
+            if (arrayOne == null || arrayTwo == null) {
+                return null;
+            }
+
+            List<Object> list = new ArrayList<>();
+            Map<ObjectContainer, Integer> map = new HashMap<>();
+            for (int pos = 0; pos < arrayOne.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayOne, pos);
+                if (element == null) {
+                    map.put(null, map.getOrDefault(null, 0) + 1);
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    map.put(objectContainer, map.getOrDefault(objectContainer, 0) + 1);
+                }
+            }
+            for (int pos = 0; pos < arrayTwo.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayTwo, pos);
+                if (element == null) {
+                    if (map.containsKey(null)) {
+                        int count = map.get(null);
+                        count--;
+                        if (count == 0) {
+                            map.remove(null);
+                        } else {
+                            map.put(null, count);
+                        }
+                    }
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    if (map.containsKey(objectContainer)) {
+                        int count = map.get(objectContainer);
+                        count--;
+                        if (count == 0) {
+                            map.remove(objectContainer);
+                        } else {
+                            map.put(objectContainer, count);
+                        }
+                    }
+                }
```

Comment:
The two branches are identical. You can do sth like:
```
final ObjectContainer objectContainer = element == null ? null : new ObjectContainer(element);
if (map.containsKey(objectContainer)) {
    int count = map.get(objectContainer);
    count--;
    if (count == 0) {
        map.remove(objectContainer);
    } else {
       map.put(objectContainer, count);
    }
}
```

Reply:
You could also use `computeIfPresent` here:

```
map.computeIfPresent(objectContainer, (k, v) -> v - 1 <= 0 ? null : v -1)
```

Reply:
ok